### PR TITLE
[SPARK-23431][CORE] Expose stage level peak executor metrics via REST API

### PR DESCRIPTION
**Pull request by imback82**
### What changes were proposed in this pull request?

Note that this PR is forked from #23340 originally written by @edwinalu.

This PR proposes to expose the peak executor metrics at the stage level via the REST APIs:
* `/applications/<application_id>/stages/`: peak values of executor metrics for **each stage** 
* `/applications/<application_id>/stages/<stage_id>/< stage_attempt_id >`: peak values of executor metrics for **each executor** for the stage, followed by peak values of executor metrics for the stage

### Why are the changes needed?

The stage level peak executor metrics can help better understand your application's resource utilization. 

### Does this PR introduce _any_ user-facing change?

1. For the `/applications/<application_id>/stages/` API, you will see the following new info for **each stage**:
```JSON
  "peakExecutorMetrics" : {
    "JVMHeapMemory" : 213367864,
    "JVMOffHeapMemory" : 189011656,
    "OnHeapExecutionMemory" : 0,
    "OffHeapExecutionMemory" : 0,
    "OnHeapStorageMemory" : 2133349,
    "OffHeapStorageMemory" : 0,
    "OnHeapUnifiedMemory" : 2133349,
    "OffHeapUnifiedMemory" : 0,
    "DirectPoolMemory" : 282024,
    "MappedPoolMemory" : 0,
    "ProcessTreeJVMVMemory" : 0,
    "ProcessTreeJVMRSSMemory" : 0,
    "ProcessTreePythonVMemory" : 0,
    "ProcessTreePythonRSSMemory" : 0,
    "ProcessTreeOtherVMemory" : 0,
    "ProcessTreeOtherRSSMemory" : 0,
    "MinorGCCount" : 13,
    "MinorGCTime" : 115,
    "MajorGCCount" : 4,
    "MajorGCTime" : 339
  }
```

2. For the `/applications/<application_id>/stages/<stage_id>/<stage_attempt_id>` API, you will see the following new info for **each executor** under `executorSummary`:
```JSON
  "peakMemoryMetrics" : {
    "JVMHeapMemory" : 0,
    "JVMOffHeapMemory" : 0,
    "OnHeapExecutionMemory" : 0,
    "OffHeapExecutionMemory" : 0,
    "OnHeapStorageMemory" : 0,
    "OffHeapStorageMemory" : 0,
    "OnHeapUnifiedMemory" : 0,
    "OffHeapUnifiedMemory" : 0,
    "DirectPoolMemory" : 0,
    "MappedPoolMemory" : 0,
    "ProcessTreeJVMVMemory" : 0,
    "ProcessTreeJVMRSSMemory" : 0,
    "ProcessTreePythonVMemory" : 0,
    "ProcessTreePythonRSSMemory" : 0,
    "ProcessTreeOtherVMemory" : 0,
    "ProcessTreeOtherRSSMemory" : 0,
    "MinorGCCount" : 0,
    "MinorGCTime" : 0,
    "MajorGCCount" : 0,
    "MajorGCTime" : 0
  }
```
, and the following at the stage level:
```JSON
"peakExecutorMetrics" : {
    "JVMHeapMemory" : 213367864,
    "JVMOffHeapMemory" : 189011656,
    "OnHeapExecutionMemory" : 0,
    "OffHeapExecutionMemory" : 0,
    "OnHeapStorageMemory" : 2133349,
    "OffHeapStorageMemory" : 0,
    "OnHeapUnifiedMemory" : 2133349,
    "OffHeapUnifiedMemory" : 0,
    "DirectPoolMemory" : 282024,
    "MappedPoolMemory" : 0,
    "ProcessTreeJVMVMemory" : 0,
    "ProcessTreeJVMRSSMemory" : 0,
    "ProcessTreePythonVMemory" : 0,
    "ProcessTreePythonRSSMemory" : 0,
    "ProcessTreeOtherVMemory" : 0,
    "ProcessTreeOtherRSSMemory" : 0,
    "MinorGCCount" : 13,
    "MinorGCTime" : 115,
    "MajorGCCount" : 4,
    "MajorGCTime" : 339
  }
```

### How was this patch tested?

Added tests.

### CODE REVIEW

**Comment by itamarst** [comment]
@imback82 Thank you!

**Comment by imback82** [comment]
retest this please

**Comment by imback82** [comment]
retest this please

**Comment by imback82** [comment]
retest this please

**Comment by imback82** [comment]
Retest this please

**Comment by imback82** [comment]
retest this please

**Comment by imback82** [comment]
retest this please

**Comment by imback82** [comment]
retest this please

**Comment by imback82** [comment]
retest this please

**Comment by imback82** [comment]
retest this please

**Comment by imback82** [comment]
retest this please

**Comment by imback82** [comment]
retest this please

**Comment by imback82** [comment]
@gengliangwang Can you help with review? Thanks in advance!

**Comment by gengliangwang** [comment]
@imback82 Thanks for the ping. I will take a look at it tomorrow.

File: core/src/test/resources/HistoryServerExpectations/minDate_app_list_json_expectation.json
Comment (gengliangwang) [inline core/src/test/resources/HistoryServerExpectations/minDate_app_list_json_expectation.json:29]: why the output order is different here?

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala
Comment (gengliangwang) [inline core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:874]: Actually I wonder why here use `maybeUpdate` instead of `update`..

Comment (imback82) [reply core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: You are right. Changed it to `update`. Thanks.

File: core/src/test/resources/HistoryServerExpectations/minDate_app_list_json_expectation.json
Comment (imback82) [reply core/src/test/resources/HistoryServerExpectations/minDate_app_list_json_expectation.json]: This is auto-generated when I ran `HistoryServerSuite.main`. I now changed to preserve the ordering to minimize the changes.

**Comment by imback82** [comment]
retest this please

**Comment by gengliangwang** [comment]
retest this please

**Comment by imback82** [comment]
retest this please

**Comment by imback82** [comment]
retest this please

**Comment by imback82** [comment]
@gengliangwang gentle ping

**Review by gengliangwang — APPROVED** [review]

**Comment by gengliangwang** [comment]
cc @srowen @vanzin

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Hi, @imback82 .
For the following [your comment](https://github.com/apache/spark/pull/23340#issuecomment-654654972), please add an empty commit with author @edwinalu . Then, Apache Spark script will make him as a co-author.
```
 I will also ask a reviewing committer to put @edwinalu as a co-author if that's the practice when a PR is "inherited". Thanks!
```

To committers, please don't merge this PR until this has a correct authorship.

**Comment by dongjoon-hyun** [comment]
To @gengliangwang, please try to recommend the authorship commit explicitly in the Apache Spark community. Otherwise, it's forgot frequently during merging steps. Thanks!

**Comment by imback82** [comment]
Thanks @dongjoon-hyun for the reminder! I added @edwinalu as a co-author with:
```
git commit --allow-empty --author="edwinalu <edwina.lu@gmail.com>" -m "Add co-author"
```

**Comment by edwinalu** [comment]
@imback82 , thanks for all your work on this!

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala
Comment (wypoon) [reply core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: I am not so familiar with `AppStatusListener`, so please educate me. From what I see, for most events, `maybeUpdate` is called. I assume that we only want to update a live application if it hasn't been updated in the last configured period. For the  `SparkListenerStageExecutorMetrics` event, according to Edwina's comment, this is only processed when reading event logs, so `update` is called in that case.

**Comment by wypoon** [comment]
In the PR description,

"This PR proposes to expose the peak executor metrics at the stage level via the REST API (/applications/<application_id>/stages/ or /applications/<application_id>/stages/<stage_id>/<stage_attemp_id>)."

there is a typo; "<stage_attemp_id>" should be "<stage_attempt_id>".

It might be helpful to clarify that the peak values of executor metrics **for the stage** are shown for each stage in the first (/applications/<application_id>/stages/) and peak values of executor metrics **for each executor for the stage** are shown for the second /applications/<application_id>/stages/<stage_id>/<stage_attempt_id>).
At least, this is my understanding of what is intended.
Edwina's PR description has

"Peak values for metrics are show for each stage. For executor summaries for each stage, the peak values per executor are also shown."

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala
Comment (imback82) [reply core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: We need `update` here since we know that new peak values are received (by checking `compareAndUpdatePeakValues`). If we use `maybeUpdate`, we may miss recording these new peak values.

**Comment by imback82** [comment]
Thanks @wypoon for the suggestion. I updated the description, and hopefully that clarifies things a bit more.

**Comment by gengliangwang** [comment]
@dongjoon-hyun Thanks for the reminder!

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala
Comment (wypoon) [reply core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: Thanks for the explanation.

**Comment by gengliangwang** [comment]
retest this please.

**Comment by gengliangwang** [comment]
@imback82 I saw this error from github action tests:
```
ERROR: this R is version 3.4.4, package 'SparkR' requires R >= 3.5
[error] running /home/runner/work/spark/spark/R/install-dev.sh
```
Could you rebase the PR to the latest master? I will merge it once the tests are passed. Thanks!

**Comment by imback82** [comment]
retest this please

**Comment by gengliangwang** [comment]
Thanks, merging to master

**Comment by dongjoon-hyun** [comment]
Thank you, @imback82 and @gengliangwang !

**Comment by imback82** [comment]
Thanks everyone!

**Pull request closed without merging**
